src/Mod/Path/PathScripts/PathDrilling.py

```diff
@@ -23,130 +23,235 @@
-
-        obj.ReturnLevel = ['G99', 'G98']  # Canned Cycle Return Level
-        obj.ExtraOffset = ['None', 'Drill Tip', '2x Drill Tip']  # Canned Cycle Return Level
+        """initCircularHoleOperation(obj) ... add drilling specific properties to obj."""
+        obj.addProperty(
+            "App::PropertyLength",
+            "PeckDepth",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property",
+                "Incremental Drill depth before retracting to clear chips",
+            ),
+        )
+        obj.addProperty(
+            "App::PropertyBool",
+            "PeckEnabled",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP("App::Property", "Enable pecking"),
+        )
+        obj.addProperty(
+            "App::PropertyFloat",
+            "DwellTime",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property", "The time to dwell between peck cycles"
+            ),
+        )
+        obj.addProperty(
+            "App::PropertyBool",
+            "DwellEnabled",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP("App::Property", "Enable dwell"),
+        )
+        obj.addProperty(
+            "App::PropertyBool",
+            "AddTipLength",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property",
+                "Calculate the tip length and subtract from final depth",
+            ),
+        )
+        obj.addProperty(
+            "App::PropertyEnumeration",
+            "ReturnLevel",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property", "Controls how tool retracts Default=G99"
+            ),
+        )
+        obj.addProperty(
+            "App::PropertyDistance",
+            "RetractHeight",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property",
+                "The height where feed starts and height during retract tool when path is finished while in a peck operation",
+            ),
+        )
+        obj.addProperty(
+            "App::PropertyEnumeration",
+            "ExtraOffset",
+            "Drill",
+            QtCore.QT_TRANSLATE_NOOP(
+                "App::Property", "How far the drill depth is extended"
+            ),
+        )
+
+        obj.ReturnLevel = ["G99", "G98"]  # Canned Cycle Return Level
+        obj.ExtraOffset = [
+            "None",
+            "Drill Tip",
+            "2x Drill Tip",
+        ]  # Canned Cycle Return Level
 
     def circularHoleExecute(self, obj, holes):
-        '''circularHoleExecute(obj, holes) ... generate drill operation for each hole in holes.'''
+        """circularHoleExecute(obj, holes) ... generate drill operation for each hole in holes."""
         PathLog.track()
+        machine = PathMachineState.MachineState()
```

Comment:
A new module and class, MachineState.  Appears to be a solution to model the user's machine state based upon the code output from the operation.  Makes sense.  A couple of things...

1.) At the moment, this `machine` object dies when the `circularHoleExecute()` method returns.  This suggests this is something to be employed in future improvements to this operation.  Yes?
 
2.) Would it not be better to move this `MachineState()` usage upstream to the `PathOp` module, and then simply do a simple for loop over the path commands after `opExecute()` returns `results` and the path command list is complete.  As stands with this implementation, there are many `machine.addCommand()` calls.  A simple loop in `PathOp.execute()` would be cleaner and also include any pre and post `opExecute()` commands added in the parent `execute()` method there.  Also, it would apply this machine state idea to all existing ops based on PathOp module.
 
3.) How is the machine state going to be updated if a user inspects the gcode for an operation, such as this drilling op, and modifies the code.  That will then potentially make machine state and the operation's gcode asychronous.  Based on this, this same machine state updating should be performed when a user manually inspects and changes the operation code.  Or, is the machine state only intended to be used during code generation, since it only lives in the `opExecute()` method?

Just some thoughts.

Reply:
This is not intended to be a state persistent representation of the machine.  It is more of a scratch pad for the operation to keep track of the machine state when generating commands internally.  That's why it's thrown away at the end.
  
Moving it upstream would make it very easy to get the final machine state that results from the operation, but it makes it useless to the operation itself.  

For example, if the operation is iterating over a bunch of drilling commands and needs to decide after each hole whether or not a rotation is necessary, it can query the machinestate to see what the current position is and then compare that position with the required position.  Without the helper, the operation needs to iterate backwards through the current commands to determine when each axis moved and to what position.

I expect this will also be useful for postprocessors.

If I were a smarter man, I'd incorporate this into the C++ of the Path object.  You could add commands to the Path or query the path to get the current machine position and rotation.